ipc/provider/src/dryrun.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: MIT
+
+//! Performs dry run for the ipc provider instead of directly submitter the txn on chain
+
+use crate::preflight::Preflight;
+use anyhow::anyhow;
+use base64::Engine;
+use ethers_contract::core::abi::{Function, Tokenizable};
+use ethers_contract::encode_function_data;
+use fvm_ipld_encoding::{BytesSer, RawBytes};
+use ipc_actors_abis::register_subnet_facet;
+use ipc_api::subnet::ConstructParams;
+use std::collections::BTreeMap;
+
+pub struct IPCDryRunProvider {
+    pub(crate) preflight: Preflight,
+}
+
+impl IPCDryRunProvider {
+    pub fn create_subnet(&self, params: ConstructParams) -> anyhow::Result<()> {
+        let params = self.preflight.create_subnet(params)?;
+        let params = register_subnet_facet::ConstructorParams::try_from(params)?;
+
+        let data = to_fvm_calldata(
+            &register_subnet_facet::REGISTERSUBNETFACET_ABI.functions,
+            "newSubnetActor",
+            params,
+        )?;
+        log::info!(
+            "params: {}",
```

Comment:
I have now re-read the ticket, and it looks like the purpose of this story is that other tools can pick up the output. 

As I have said multiple times, there is a difference between logging and output. Logging should go to _stderr_ and purposeful output to _stdout_, so that you can pipe it to the next tool without being conflated with logs, or having to parse the thing out from among timestamps, or as here, English prefix. 

For this purpose I always recommended using `println!` with `json!`, like in `fendermint`, which is human readable as well as easily parseable with `jq`. A Base64 string in the logs is not human readable and I imagine difficult to parse as well.

Reply:
I'm not sure if you have seen https://linear.app/interplanetary-consensus/issue/ENG-780/implement-dry-run-for-ipc-cli-commands#comment-b4ad2e7d about the list of fields that @raulk thought this tool should print. 

This is where it's not entirely clear what `--dry-run` was supposed to be. Just thinking about it, I would have guessed that it would print out what it would send to the ethereum API, something you could copy and use with `curl` for example, so you have a sense of everything, the destination, the sending user, etc. 

Instead, the ticket is talking about the multisig, printing stuff that can be signed and sent to the `Multisig` actor, which means an FVM `Message` needs to be constructed, which the `ipc-cli` doesn't do, although the `params` constructed here would be part of it. However, as we discussed it would potentially be helpful for them to know what the other knowable fields wouldbe: the `to`, the `method_number`, `value`; we can print `from` to be this signer, even though it's useless for multisig, just to keep it general. 

So, just wondering if this field is enough, as it keeps the `ipc-cli` from having to construct a `Message`, but (probably) still makes the use of `msig` possible, or do you want it to list more fields.

Reply:
OTOH you can forget what I said about doing this on the HTTP level as it would print a signed transaction request, which isn't what is ultimately being sent to the EVM actor. Although maybe it _would_ make sense to detect the JSON-RPC request going to `eth_sendRawTransaction` and just replicate [these lines](https://github.com/consensus-shipyard/ipc/blob/ae750fe9778d2b6dda10768c3b6966ba3c648238/fendermint/eth/api/src/apis/eth.rs#L606-L621) and print every field from the `Message`.

Reply:
Maybe this is what I'd do: instead of `--dry-run` begin an on/off flag, it could be a `--dry-run=[message|raw]` format selector, which, if set, would result in the `eth_sendRawTransaction` begin hijacked and the transaction printed either as raw RLP, or as parsed `Message` fields in JSON. 

This would allow all other calls such as fee estimation and nonce checking to go through, so you would get a fully configured transaction to print.

The only problem with it in our case is that `eth_call` and `eth_estimateGas` would fail because they actually execute the transaction, and the caller would not be authorized to actually set the federated powers, only the multisig would, so I assume we'd never get to the point of actually sending anything. You would have to prevent gas estimation by setting some dummy values. But that can be achieved, because the `ipc-cli` could know that when it's asked to set the federated powers, to avoid triggering an estimation.

Reply:
I'm not sure if we want to perform any nonce check because `ethers` should have checked for us and dry-run, mostly should be just print out what you are sending?

I think `--dry-run=[message|raw]` works for me, but curious why not set it as a separate field?

Reply:
btw, when I'm implementing it, I feel `raw` somehow does not make sense without the context, like `evm` or `fvm`, I kind of feel it's two params, `network` and `format`. Without the network, I dont really know what does `raw` mean. I kind of feel if we know the network, i.e. `dry-run=[evm|fvm]` and prints the fields of the txn, it should be enough for now?

Reply:
Yeah, `evm|fvm` would work great :+1:

Reply:
Or, well, maybe not. You see the `fvm` is a given, we always send a `Message` type at the end. But `evm` is specific to subnets: some can be of the `fevm` type (which is the only one supported right now), but you have a `Provider` abstraction in case it's not. That's what I meant by `raw`: print out what you would send to the API that corresponds to the subnet. I don't want an non-evm provider to produce evm format, only it's own, or potentially `fvm`.

Reply:
I'm just trying to make sense of how to interpret this whole `--dry-run` in a generic way. Obviously every EVM request can be printed, which, as you say, "should be just print out what you are sending", however the FVM version is _not_ what _you_ are sending, it's something extra, but it is what Raul was actually asking for, but it doesn't sit well with `--dry-run` as a general concept IMO. That's why I thought `--dry-run=fvm` can give the hint. 

Whether it's one fields or not: if it would be two fields, then `--dry-run` could be the default "API native" format, ie. EVM in this case, and you could use a `--dry-run-format=fvm` to differ from that.